// databaker/lib/chart/bar-chart.js
"use strict";

require("core-js/modules/es.array.iterator");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BarChart = void 0;

var _canvas = require("canvas");

var _d3Scale = require("d3-scale");

var _canvasImage = require("../image/canvas-image");

var _utils = require("./utils");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class BarChart {
  constructor(options = {}) {
    _defineProperty(this, "options", void 0);

    this.options = {
      borderColor: options.borderColor || "#FFFFFF",
      textColor: options.textColor || "#FFFFFF",
      barColors: options.barColors && options.barColors.length ? options.barColors : ["#0074D9", "#B10DC9", "#7FDBFF", "#F012BE", "#39CCCC", "#85144b", "#3D9970", "#FF4136", "#2ECC40", "#FF851B", "#01FF70", "#FFDC00"],
      width: options.width || 800,
      height: options.height || 400
    };
  }

  draw(labels, datas) {
    const canvas = (0, _canvas.createCanvas)(this.options.width, this.options.height);
    const margin = {
      top: 20,
      right: 20,
      bottom: 30,
      left: 40
    };
    const width = canvas.width - margin.left - margin.right;
    const height = canvas.height - margin.top - margin.bottom;
    const x = (0, _d3Scale.scaleBand)().padding(0.1).rangeRound([0, width]).domain(labels);
    const yMin = Math.min(...datas.map(d => Math.min(...d)));
    const yMax = Math.max(...datas.map(d => Math.max(...d)));
    const y = (0, _d3Scale.scaleLinear)().rangeRound([height, 0]).domain([yMin - (yMax - yMin) * 0.03, yMax]); // draw

    const context = canvas.getContext("2d");
    context.translate(margin.left, margin.top); // draw axis

    (0, _utils.drawAxis)(context, width, height, x, y, this.options);
    const bar = (0, _d3Scale.scaleBand)().padding(0.2).rangeRound([0, x.bandwidth()]).domain(Object.keys(datas));
    datas.forEach((data, dataIndex) => {
      context.fillStyle = this.options.barColors[dataIndex % this.options.barColors.length];
      labels.forEach((label, labelIndex) => {
        if (data[labelIndex]) {
          context.fillRect(x(label) + bar(dataIndex + ""), y(data[labelIndex]), bar.bandwidth(), height - y(data[labelIndex]));
        }
      });
    });
    return new _canvasImage.CanvasImage(canvas);
  }

}

exports.BarChart = BarChart;
